File: cpp/benchmark.cpp
// Benchmark 10 numerical operations using Eigen
#include <Eigen/Dense>
#include <Eigen/QR>
#include <Eigen/LU>
#include <Eigen/Cholesky>
#include <Eigen/SVD>
#include <Eigen/Eigenvalues>
#include <unsupported/Eigen/FFT>
#include <algorithm>

#include <chrono>
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <vector>

using namespace Eigen;
using Clock = std::chrono::high_resolution_clock;

// ── Anti-optimisation sink ─────────────────────────────────────────────────────
// Prevents the compiler from eliding computations whose results aren't used.
static volatile double g_sink = 0.0;
template<typename T>
void do_not_optimise(const T& val) {
    // Touch one element so the whole computation is retained.
    if constexpr (std::is_arithmetic_v<T>) {
        g_sink += (double)val;
    } else {
        g_sink += val.sum();   // works for Eigen matrices/vectors
    }
}

// ── RNG / matrix builders ─────────────────────────────────────────────────────

static constexpr unsigned SEED = 42;

MatrixXd rand_matrix(int n, int m) {
    std::mt19937_64 rng(SEED);
    std::normal_distribution<double> dist(0.0, 1.0);
    MatrixXd A(n, m);
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < m; ++j)
            A(i, j) = dist(rng);
    return A;
}

VectorXd rand_vec(int n) {
    std::mt19937_64 rng(SEED + 1);
    std::normal_distribution<double> dist(0.0, 1.0);
    VectorXd v(n);
    for (int i = 0; i < n; ++i) v(i) = dist(rng);
    return v;
}

MatrixXd make_spd(int n) {
    MatrixXd A = rand_matrix(n, n);
    return A * A.transpose() + (double)n * MatrixXd::Identity(n, n);
}

MatrixXd make_sym(int n) {
    MatrixXd A = rand_matrix(n, n);
    return (A + A.transpose()) * 0.5;
}

// ── Simple JSON helpers ───────────────────────────────────────────────────────

static std::string jstr(const std::string& s)  { return "\"" + s + "\""; }
static std::string jkv(const std::string& k, double v) {
    std::ostringstream o; o << "\"" << k << "\": " << v; return o.str();
}
static std::string jkvs(const std::string& k, const std::string& v) {
    return "\"" + k + "\": \"" + v + "\"";
}

// ── Memory helpers ────────────────────────────────────────────────────────────
// Eigen uses malloc() directly, so operator new tracking can't capture it.
// We compute output allocation size to keep the metric comparable with other
// languages (Python tracemalloc, Julia @allocated, R gc() — all capture the
// output matrix allocation and any intermediates).
static double mb(size_t bytes) { return (double)bytes / (1024.0 * 1024.0); }
static double mat_mb(int n, int m = -1) {
    if (m < 0) m = n;
    return mb((size_t)n * m * sizeof(double));
}
static double vec_mb(int n) { return mb((size_t)n * sizeof(double)); }

// ── Benchmark harness ─────────────────────────────────────────────────────────

static constexpr int N_WARMUP = 3;
static constexpr int N_RUNS   = 10;

struct OpResult {
    std::string name, description;
    double mean_ms, std_ms, min_ms, memory_mb;
};

template<typename Setup, typename Op>
OpResult bench(const std::string& name, const std::string& desc,
               double output_memory_mb, Setup setup, Op op) {
    for (int i = 0; i < N_WARMUP; ++i) { auto d = setup(); op(d); }

    std::vector<double> times_ms(N_RUNS);

    for (int i = 0; i < N_RUNS; ++i) {
        auto data = setup();
        auto t0 = Clock::now();
        op(data);
        auto t1 = Clock::now();
        times_ms[i] = std::chrono::duration<double, std::milli>(t1 - t0).count();
    }

    double sum = 0; for (double t : times_ms) sum += t;
    double mean = sum / N_RUNS;
    double var  = 0; for (double t : times_ms) var += (t - mean)*(t - mean);
    double mn   = times_ms[0]; for (double t : times_ms) mn = std::min(mn, t);

    std::printf("  Benchmarking: %-25s ... %.2f ms  (%.2f MB)\n",
                name.c_str(), mean, output_memory_mb);

    return { name, desc, mean, std::sqrt(var / N_RUNS), mn, output_memory_mb };
}

// ── Main ──────────────────────────────────────────────────────────────────────

int main() {
    std::string version = "C++17 / Eigen " +
        std::to_string(EIGEN_WORLD_VERSION) + "." +
        std::to_string(EIGEN_MAJOR_VERSION) + "." +
        std::to_string(EIGEN_MINOR_VERSION);
    std::cout << version << "\n";

    std::vector<OpResult> results;

    // 1. Matrix Multiply — result: 1000×1000
    results.push_back(bench(
        "Matrix Multiply", "1000×1000 dense matrix multiplication (A * B)",
        mat_mb(1000),
        []{ return std::make_pair(rand_matrix(1000,1000), rand_matrix(1000,1000)); },
        [](auto& d){ MatrixXd c = d.first * d.second; do_not_optimise(c); }
    ));

    // 2. Matrix Inverse — result: 500×500
    results.push_back(bench(
        "Matrix Inverse", "Inversion of 500×500 SPD matrix",
        mat_mb(500),
        []{ return make_spd(500); },
        [](auto& a){ MatrixXd inv = a.inverse(); do_not_optimise(inv); }
    ));

    // 3. LU — stores L+U in 500×500 matrix
    results.push_back(bench(
        "LU Decomposition", "LU factorisation of 500×500 matrix",
        mat_mb(500),
        []{ return rand_matrix(500, 500); },
        [](auto& a){ PartialPivLU<MatrixXd> lu(a); do_not_optimise(lu.matrixLU()); }
    ));

    // 4. Eigenvalue — eigenvectors 300×300 + eigenvalues 300
    results.push_back(bench(
        "Eigenvalue Decomp", "Full eigendecomposition of 300×300 symmetric matrix",
        mat_mb(300) + vec_mb(300),
        []{ return make_sym(300); },
        [](auto& a){ SelfAdjointEigenSolver<MatrixXd> es(a); do_not_optimise(es.eigenvalues()); }
    ));

    // 5. Cholesky — lower triangle 500×500
    results.push_back(bench(
        "Cholesky", "Cholesky factorisation of 500×500 SPD matrix",
        mat_mb(500),
        []{ return make_spd(500); },
        [](auto& a){ LLT<MatrixXd> llt(a); do_not_optimise(llt.matrixL().toDenseMatrix()); }
    ));

    // 6. SVD — U:500×300, S:300, Vt:300×300
    results.push_back(bench(
        "SVD", "Economy SVD of 500×300 matrix (ComputeThinU|ComputeThinV; U:500×300)",
        mat_mb(500,300) + vec_mb(300) + mat_mb(300),
        []{ return rand_matrix(500, 300); },
        [](auto& a){ BDCSVD<MatrixXd> svd(a, ComputeThinU | ComputeThinV); do_not_optimise(svd.singularValues()); }
    ));

    // 7. Linear Solve — result: vector 1000
    results.push_back(bench(
        "Linear System Solve", "Solve Ax=b for 1000×1000 A, 1000 b",
        vec_mb(1000),
        []{ return std::make_pair(make_spd(1000), rand_vec(1000)); },
        [](auto& d){ VectorXd x = d.first.llt().solve(d.second); do_not_optimise(x); }
    ));

    // 8. Vector Dot Product — result: scalar
    results.push_back(bench(
        "Vector Dot Product", "Dot product of two 10M-element vectors",
        vec_mb(1),
        []{ return std::make_pair(rand_vec(10'000'000), rand_vec(10'000'000)); },
        [](auto& d){ double v = d.first.dot(d.second); do_not_optimise(v); }
    ));

    // 9. Hadamard — result: 1000×1000
    results.push_back(bench(
        "Hadamard Product", "Element-wise multiply + add on 1000×1000 matrices",
        mat_mb(1000),
        []{ return std::make_tuple(rand_matrix(1000,1000), rand_matrix(1000,1000), rand_matrix(1000,1000)); },
        [](auto& d){
            MatrixXd r = std::get<0>(d).cwiseProduct(std::get<1>(d)) + std::get<2>(d);
            do_not_optimise(r);
        }
    ));

    // 10. QR — Q:500×500 + R:500×500
    results.push_back(bench(
        "QR Decomposition", "QR factorisation of 500×500 matrix",
        mat_mb(500),
        []{ return rand_matrix(500, 500); },
        [](auto& a){ HouseholderQR<MatrixXd> qr(a); do_not_optimise(qr.matrixQR()); }
    ));

    // 11. FFT (real, 1M)
    results.push_back(bench(
        "FFT (real, 1M)", "Real FFT of 2^20=1M-element vector (Eigen::FFT / KissFFT)",
        (vec_mb(1 << 20) / 2.0 + vec_mb(1)) * 2,  // ~complex output size
        []{ return rand_vec(1 << 20); },
        [](auto& v) {
            Eigen::FFT<double> fft;
            std::vector<std::complex<double>> out;
            std::vector<double> in(v.data(), v.data() + v.size());
            fft.fwd(out, in);
            do_not_optimise(out[0].real());
        }
    ));

    // 12. Sort 10M floats
    results.push_back(bench(
        "Sort 10M floats", "Unstable sort of 10M random float64 values (std::sort)",
        0.0,
        []{ return rand_vec(10'000'000); },
        [](auto& v) {
            std::vector<double> data(v.data(), v.data() + v.size());
            std::sort(data.begin(), data.end());
            do_not_optimise(data[0]);
        }
    ));

    // ── Write JSON ────────────────────────────────────────────────────────────
    std::filesystem::create_directories("results");
    std::ofstream out("results/cpp_results.json");
    out << "{\n";
    out << "  " << jkvs("language", "C++") << ",\n";
    out << "  " << jkvs("version", version) << ",\n";
    out << "  " << jkvs("platform", "native") << ",\n";
    out << "  \"operations\": [\n";
    for (size_t i = 0; i < results.size(); ++i) {
        const auto& r = results[i];
        out << "    {\n";
        out << "      " << jkvs("name", r.name) << ",\n";
        out << "      " << jkvs("description", r.description) << ",\n";
        out << "      " << jkv("mean_ms", r.mean_ms) << ",\n";
        out << "      " << jkv("std_ms", r.std_ms) << ",\n";
        out << "      " << jkv("min_ms", r.min_ms) << ",\n";
        out << "      " << jkv("memory_mb", r.memory_mb) << "\n";
        out << "    }" << (i + 1 < results.size() ? "," : "") << "\n";
    }
    out << "  ]\n}\n";
    out.close();
    std::cout << "Saved results/cpp_results.json\n";
    return 0;
}

File: octave/benchmark.m
% octave/benchmark.m
% Run from project root: octave --no-gui octave/benchmark.m

N_WARMUP = 3;
N_RUNS   = 10;

% Fixed seed for reproducibility
randn ('state', 42);

ver_str = version ();
printf ('Octave %s\n', ver_str);

results = {};

%% ── 1. Matrix Multiply ──────────────────────────────────────────────────────
name = 'Matrix Multiply';
desc = '1000x1000 dense matrix multiplication (A * B)';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  A = randn (1000, 1000); B = randn (1000, 1000); C = A * B;
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  A = randn (1000, 1000); B = randn (1000, 1000);
  tic; C = A * B; times(k) = toc () * 1000;
end
mem_mb = (1000 * 1000 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 2. Matrix Inverse ───────────────────────────────────────────────────────
name = 'Matrix Inverse';
desc = 'Inversion of 500x500 SPD matrix';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  B = randn (500, 500); A = B * B' + 500 * eye (500); C = inv (A);
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  B = randn (500, 500); A = B * B' + 500 * eye (500);
  tic; C = inv (A); times(k) = toc () * 1000;
end
mem_mb = (500 * 500 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 3. LU Decomposition ─────────────────────────────────────────────────────
name = 'LU Decomposition';
desc = 'LU factorisation of 500x500 matrix';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  A = randn (500, 500); [L, U, P] = lu (A);
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  A = randn (500, 500);
  tic; [L, U, P] = lu (A); times(k) = toc () * 1000;
end
mem_mb = (500 * 500 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 4. Eigenvalue Decomposition ─────────────────────────────────────────────
name = 'Eigenvalue Decomp';
desc = 'Full eigendecomposition of 300x300 symmetric matrix';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  A = randn (300, 300); A = (A + A') / 2; [V, D] = eig (A);
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  A = randn (300, 300); A = (A + A') / 2;
  tic; [V, D] = eig (A); times(k) = toc () * 1000;
end
mem_mb = (300 * 300 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 5. Cholesky ─────────────────────────────────────────────────────────────
name = 'Cholesky';
desc = 'Cholesky factorisation of 500x500 SPD matrix';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  B = randn (500, 500); A = B * B' + 500 * eye (500); L = chol (A, 'lower');
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  B = randn (500, 500); A = B * B' + 500 * eye (500);
  tic; L = chol (A, 'lower'); times(k) = toc () * 1000;
end
mem_mb = (500 * 500 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 6. SVD ──────────────────────────────────────────────────────────────────
name = 'SVD';
desc = 'Economy SVD of 500x300 matrix (U:500x300, S:300, V:300x300)';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  A = randn (500, 300); [U, S, V] = svd (A, 'econ');
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  A = randn (500, 300);
  tic; [U, S, V] = svd (A, 'econ'); times(k) = toc () * 1000;
end
mem_mb = (300 * 300 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 7. Linear System Solve ──────────────────────────────────────────────────
name = 'Linear System Solve';
desc = 'Solve Ax=b for 1000x1000 A, 1000 b';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  B = randn (1000, 1000); A = B * B' + 1000 * eye (1000); b = randn (1000, 1);
  x = A \ b;
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  B = randn (1000, 1000); A = B * B' + 1000 * eye (1000); b = randn (1000, 1);
  tic; x = A \ b; times(k) = toc () * 1000;
end
mem_mb = (1000 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 8. Vector Dot Product ───────────────────────────────────────────────────
name = 'Vector Dot Product';
desc = 'Dot product of two 10M-element vectors';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  a = randn (10000000, 1); b = randn (10000000, 1); s = dot (a, b);
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  a = randn (10000000, 1); b = randn (10000000, 1);
  tic; s = dot (a, b); times(k) = toc () * 1000;
end
mem_mb = 8 / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 9. Hadamard Product ─────────────────────────────────────────────────────
name = 'Hadamard Product';
desc = 'Element-wise multiply + add on 1000x1000 matrices';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  A = randn (1000, 1000); B = randn (1000, 1000); C = randn (1000, 1000);
  D = A .* B + C;
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  A = randn (1000, 1000); B = randn (1000, 1000); C = randn (1000, 1000);
  tic; D = A .* B + C; times(k) = toc () * 1000;
end
mem_mb = (1000 * 1000 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 10. QR Decomposition ────────────────────────────────────────────────────
name = 'QR Decomposition';
desc = 'QR factorisation of 500x500 matrix';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  A = randn (500, 500); [Q, R] = qr (A);
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  A = randn (500, 500);
  tic; [Q, R] = qr (A); times(k) = toc () * 1000;
end
mem_mb = (500 * 500 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 11. FFT ─────────────────────────────────────────────────────────────────
name = 'FFT (real, 1M)';
desc = 'Full complex FFT sliced to N/2+1 (Octave fft; no native rfft)';
printf ('  Benchmarking: %s ...', name);
N_FFT = 2^20;
for k = 1:N_WARMUP
  v = randn (N_FFT, 1); f = fft (v); f = f(1:N_FFT/2+1);
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  v = randn (N_FFT, 1);
  tic; f = fft (v); f = f(1:N_FFT/2+1); times(k) = toc () * 1000;
end
mem_mb = ((N_FFT / 2 + 1) * 16) / 1024^2;  % complex doubles
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── 12. Sort ────────────────────────────────────────────────────────────────
name = 'Sort 10M floats';
desc = 'Sort of 10M random float64 values';
printf ('  Benchmarking: %s ...', name);
for k = 1:N_WARMUP
  v = randn (10000000, 1); s = sort (v);
end
times = zeros (1, N_RUNS);
for k = 1:N_RUNS
  v = randn (10000000, 1);
  tic; s = sort (v); times(k) = toc () * 1000;
end
mem_mb = (10000000 * 8) / 1024^2;
results{end+1} = struct ('name', name, 'description', desc, ...
  'mean_ms', mean (times), 'std_ms', std (times), 'min_ms', min (times), 'memory_mb', mem_mb);
printf (' %.2f ms  (%.2f MB)\n', mean (times), mem_mb);

%% ── Write JSON ──────────────────────────────────────────────────────────────
% Build operations array as JSON string
ops_json = '[';
for k = 1:length (results)
  r = results{k};
  name_esc = strrep (r.name, '"', '\"');
  desc_esc = strrep (r.description, '"', '\"');
  entry = sprintf ('{"name":"%s","description":"%s","mean_ms":%.6f,"std_ms":%.6f,"min_ms":%.6f,"memory_mb":%.6f}', ...
    name_esc, desc_esc, r.mean_ms, r.std_ms, r.min_ms, r.memory_mb);
  if k < length (results)
    ops_json = [ops_json, entry, ','];
  else
    ops_json = [ops_json, entry];
  end
end
ops_json = [ops_json, ']'];

platform_str = computer ();
out = sprintf ('{\n  "language": "Octave",\n  "version": "Octave %s",\n  "platform": "%s",\n  "operations": %s\n}\n', ...
  ver_str, platform_str, ops_json);

[status, ~] = mkdir ('results');
fid = fopen ('results/octave_results.json', 'w');
fputs (fid, out);
fclose (fid);
printf ('Saved results/octave_results.json\n');
